File: packages/ik_net/test/ranking_update_test.dart
import 'package:ik_net/ik_net.dart';
import 'package:test/test.dart';

void main() {
  const noon = 1786568400000; // 2026-08-13T00:00:00.000Z? pinned like other net tests
  final day = DateTime.fromMillisecondsSinceEpoch(noon, isUtc: true);
  final startOfDay = DateTime.utc(day.year, day.month, day.day).millisecondsSinceEpoch;
  final nextDay = DateTime.utc(day.year, day.month, day.day).add(const Duration(days: 1)).millisecondsSinceEpoch;

  test('names the storage key after the account', () {
    expect(rankingUpdateStorageKey('usr_0001'), 'idle-kingdoms.leaderboard.submit-at:usr_0001');
  });

  test('auto-submits when nothing has been posted, or the UTC day rolled over', () {
    expect(shouldAutoSubmitRanking(null, startOfDay), isTrue);
    expect(shouldAutoSubmitRanking(startOfDay, startOfDay + 60 * 60 * 1000), isFalse);
    expect(shouldAutoSubmitRanking(startOfDay, nextDay), isTrue);
    expect(utcDayKey(startOfDay), utcDayKey(startOfDay + 12 * 60 * 60 * 1000));
    expect(utcDayKey(startOfDay), isNot(utcDayKey(nextDay)));
  });

  test('manual updates wait an hour', () {
    expect(canUpdateRanking(null, startOfDay), isTrue);
    expect(canUpdateRanking(startOfDay, startOfDay + rankingUpdateCooldownMs - 1), isFalse);
    expect(canUpdateRanking(startOfDay, startOfDay + rankingUpdateCooldownMs), isTrue);
    expect(rankingCooldownRemainingMs(startOfDay, startOfDay + 15 * 60 * 1000), 45 * 60 * 1000);
    expect(
      rankingCooldownMessage(45 * 60 * 1000),
      'You can update your ranking again in 45 minutes.',
    );
    expect(rankingCooldownMessage(60 * 1000), 'You can update your ranking again in 1 minute.');
    expect(rankingCooldownMessage(rankingUpdateCooldownMs), 'You can update your ranking again in 1 hour.');
    expect(rankingUpdateHint(startOfDay, startOfDay), 'You can update your ranking again in 1 hour.');
    expect(rankingUpdateHint(null, startOfDay), rankingUpdateReadyHint);
  });

  test('parses a stored timestamp', () {
    expect(parseRankingSubmitAt(null), isNull);
    expect(parseRankingSubmitAt(''), isNull);
    expect(parseRankingSubmitAt('1786568400000'), 1786568400000);
  });
}

File: src/game/multiplayer/rankingUpdate.test.ts
import { describe, expect, it } from 'vitest'
import {
  canUpdateRanking,
  msUntilNextUtcHour,
  parseRankingSubmitAt,
  rankingCooldownMessage,
  rankingCooldownRemainingMs,
  RANKING_PUBLISH_DEBOUNCE_MS,
  RANKING_UPDATE_COOLDOWN_MS,
  RANKING_UPDATE_READY_HINT,
  rankingUpdateHint,
  rankingUpdateStorageKey,
  shouldAutoSubmitRanking,
  shouldPublishForUtcHour,
  shouldPublishOnOpen,
  utcDayKey,
  utcHourKey,
} from './rankingUpdate'

const noon = 1_786_568_400_000
const startOfDay = Date.UTC(
  new Date(noon).getUTCFullYear(),
  new Date(noon).getUTCMonth(),
  new Date(noon).getUTCDate(),
)
const nextDay = startOfDay + 24 * 60 * 60 * 1000

describe('ranking update policy', () => {
  it('names the storage key after the account', () => {
    expect(rankingUpdateStorageKey('usr_0001')).toBe('idle-kingdoms.leaderboard.submit-at:usr_0001')
  })

  it('auto-submits when nothing has been posted, or the UTC day rolled over', () => {
    expect(shouldAutoSubmitRanking(null, startOfDay)).toBe(true)
    expect(shouldAutoSubmitRanking(startOfDay, startOfDay + 60 * 60 * 1000)).toBe(false)
    expect(shouldAutoSubmitRanking(startOfDay, nextDay)).toBe(true)
    expect(utcDayKey(startOfDay)).toBe(utcDayKey(startOfDay + 12 * 60 * 60 * 1000))
    expect(utcDayKey(startOfDay)).not.toBe(utcDayKey(nextDay))
  })

  it('waits an hour between manual updates', () => {
    expect(canUpdateRanking(null, startOfDay)).toBe(true)
    expect(canUpdateRanking(startOfDay, startOfDay + RANKING_UPDATE_COOLDOWN_MS - 1)).toBe(false)
    expect(canUpdateRanking(startOfDay, startOfDay + RANKING_UPDATE_COOLDOWN_MS)).toBe(true)
    expect(rankingCooldownRemainingMs(startOfDay, startOfDay + 15 * 60 * 1000)).toBe(45 * 60 * 1000)
    expect(rankingCooldownMessage(45 * 60 * 1000)).toBe('You can update your ranking again in 45 minutes.')
    expect(rankingCooldownMessage(60 * 1000)).toBe('You can update your ranking again in 1 minute.')
    expect(rankingCooldownMessage(RANKING_UPDATE_COOLDOWN_MS)).toBe(
      'You can update your ranking again in 1 hour.',
    )
    expect(rankingUpdateHint(startOfDay, startOfDay)).toBe('You can update your ranking again in 1 hour.')
    expect(rankingUpdateHint(null, startOfDay)).toBe(RANKING_UPDATE_READY_HINT)
  })

  it('uses UTC hour marks and a short debounce for opening the app', () => {
    expect(utcHourKey(startOfDay).endsWith('T00')).toBe(true)
    expect(utcHourKey(startOfDay + 15 * 60 * 1000)).toBe(utcHourKey(startOfDay))
    expect(msUntilNextUtcHour(startOfDay)).toBe(60 * 60 * 1000)
    expect(msUntilNextUtcHour(startOfDay + 15 * 60 * 1000)).toBe(45 * 60 * 1000)
    expect(shouldPublishForUtcHour(null, startOfDay)).toBe(true)
    expect(shouldPublishForUtcHour(startOfDay, startOfDay + 59 * 60 * 1000)).toBe(false)
    expect(shouldPublishForUtcHour(startOfDay, startOfDay + 60 * 60 * 1000)).toBe(true)
    expect(shouldPublishOnOpen(null, startOfDay)).toBe(true)
    expect(shouldPublishOnOpen(startOfDay, startOfDay + 1000)).toBe(false)
    expect(shouldPublishOnOpen(startOfDay, startOfDay + RANKING_PUBLISH_DEBOUNCE_MS)).toBe(true)
  })

  it('parses a stored timestamp', () => {
    expect(parseRankingSubmitAt(null)).toBeNull()
    expect(parseRankingSubmitAt('')).toBeNull()
    expect(parseRankingSubmitAt('1786568400000')).toBe(1786568400000)
  })
})
